### PR TITLE
[FLINK-31838][runtime] Moves thread handling for leader event listener calls from DefaultMultipleComponentLeaderElectionService into DefaultLeaderElectionService

### PR DESCRIPTION
**Pull request by XComp**
PR order:
- https://github.com/apache/flink/pull/21742
- https://github.com/apache/flink/pull/22379
- === this PR === FLINK-31838
- https://github.com/apache/flink/pull/22380
- https://github.com/apache/flink/pull/22384
- https://github.com/apache/flink/pull/22390 (FLINK-31785/FLINK-31786)
- https://github.com/apache/flink/pull/22404

## What is the purpose of the change

We need to handle the leader event calls in a separate thread in `DefaultLeaderElectionService` to ensure that they are not called in the main thread when registering the `LeaderContender` in the `LeaderElectionService`. The legacy implementations of `LeaderElectionDriver` already do this implicitly because their leader election implementations run in a dedicated thread.

The MultipleComponentLeaderElection framework had to deal with this already. I was implemented in `DefaultMultipleComponentLeaderElectionService` where a singleThread executor service was utilized to ensure the order of grant and revoke calls.

## Brief change log

* Renames `LeaderElectionEventHandler` methods to reflect the async nature
* Introduced additional asynchronous methods in `LeaderElectionEventHandler` that enable the legacy `LeaderElectionDriver` implementations to not run asynchronously. As mentioned before: These implementations have their event calls already being executed in the implementations' dedicated event threads.
* Moved `leadershipOperationExecutor` from `DefaultMultipleComponentLeaderElectionService` into `DefaultLeaderElectionService`

## Verifying this change

* Added tests for non-blocking operations to `DefaultLeaderElectionTest`
* Existing tests were refactored and should still pass

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: yes
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

### CODE REVIEW

**Comment by XComp** [comment]
@flinkbot run azure

File: flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/LeaderElectionEventHandler.java
Comment (reswqa) [inline flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/LeaderElectionEventHandler.java:49]: I'm a bit confused, why is this method named with async suffix? I know this is to make `DefaultMultipleComponentLeaderElectionService` use the executor in `DefaultLeaderElectionService`, but from the protocol of the interface, it's a bit difficult to understand. At least we can emphasize in Java doc that this method must not be executed on the caller thread.

The following is the doc for the `thenRunAsync` method of `CompletionStage`, which looks much more intuitive.
```
    /*
     * Returns a new CompletionStage that, when this stage completes
     * normally, executes the given action using this stage's default
     * asynchronous execution facility.
     */
    public CompletionStage<Void> thenRunAsync(Runnable action);
```

File: flink-kubernetes/src/main/java/org/apache/flink/kubernetes/highavailability/KubernetesLeaderElectionDriver.java
Comment (reswqa) [inline flink-kubernetes/src/main/java/org/apache/flink/kubernetes/highavailability/KubernetesLeaderElectionDriver.java:200]: Can `Executors.directExecutor()` be used here to replace `Executors.newDirectExecutorService()` 🤔

**Review by reswqa — COMMENTED** [review]
Thanks @XComp for creating this, I have left some comments, PTAL.

File: flink-runtime/src/test/java/org/apache/flink/runtime/leaderelection/DefaultLeaderElectionServiceTest.java
Comment (reswqa) [inline flink-runtime/src/test/java/org/apache/flink/runtime/leaderelection/DefaultLeaderElectionServiceTest.java:446]: To be honest, I didn't understand the intention of this test. If we run it in a synchronized manner, it shouldn't timeout either.

File: flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/LeaderElectionEventHandler.java
Comment (XComp) [reply flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/LeaderElectionEventHandler.java:49]: You're right - iterating over this change once more makes me realize that it's a bad proposal. The directExecutor doesn't need to be specified through a parameter. Instead, we could keep `onGrantLeadership(UUID)` (analogously for the other methods) to call the logic synchronously. This can be used by the legacy `LeaderElectionDriver` implementations.

The asynchronous call that's going to be triggered from the `DefaultMultipleComponentLeaderElectionService` could be named something like `forwardGrantLeadershipEvent(UUID)` (analogously for the other methods). In its JavaDoc we have to define that this method is expected to run in a separate thread.

I'm gonna go ahead a work on a commit adjusting the interface tomorrow.

File: flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/DefaultLeaderElectionService.java
Comment (XComp) [reply flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/DefaultLeaderElectionService.java:76]: my bad: that's gonna go be removed again. :+1:

File: flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/DefaultLeaderElectionService.java
Comment (XComp) [reply flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/DefaultLeaderElectionService.java:78]: copy and paste error: I'm going to move it to the other constructor and reduce the visibility on that one.

File: flink-kubernetes/src/main/java/org/apache/flink/kubernetes/highavailability/KubernetesLeaderElectionDriver.java
Comment (XComp) [reply flink-kubernetes/src/main/java/org/apache/flink/kubernetes/highavailability/KubernetesLeaderElectionDriver.java:200]: Yes, good point. I'm gonna switch

File: flink-runtime/src/test/java/org/apache/flink/runtime/leaderelection/DefaultLeaderElectionServiceTest.java
Comment (XComp) [reply flink-runtime/src/test/java/org/apache/flink/runtime/leaderelection/DefaultLeaderElectionServiceTest.java:453]: You're correct. Yesterday wasn't my strongest performance in that regards :innocent: I updated the tests.

File: flink-runtime/src/test/java/org/apache/flink/runtime/leaderelection/TestingLeaderElectionEventHandler.java
Comment (reswqa) [inline flink-runtime/src/test/java/org/apache/flink/runtime/leaderelection/TestingLeaderElectionEventHandler.java:79]: How about marking `TestingLeaderElectionEventHandler` and `TestingLeaderContender` as `@Deprecated` first to prevent other developers from using this class for testing and reduce the workload of ultimately removing these two classes. 🤔

**Review by reswqa — COMMENTED** [review]
Thanks @XComp for the update. This change looks good to me overall, I only left some minor comments. 

Let's squash some fix-up commits and wait for @zentol's another round review.

File: flink-runtime/src/test/java/org/apache/flink/runtime/leaderelection/TestingLeaderElectionEventHandler.java
Comment (XComp) [reply flink-runtime/src/test/java/org/apache/flink/runtime/leaderelection/TestingLeaderElectionEventHandler.java:79]: I'm also not happy about having all the different testing implementations. The two you mentioned have actually a purpose, though. They implement basic functionality that would be annoying to implement for every test case using the new `TestingGeneric*` implementations. Therefore, I would suggest leaving them as is.

**Comment by XComp** [comment]
Thanks @reswqa for your comments. I addressed them and push the commits. Looks like CI will remain green as well. Therefore, the PR is ready to be reviewed. I would go ahead and squash everything into reasonable chunks after the review is done.

File: flink-runtime/src/test/java/org/apache/flink/runtime/leaderelection/TestingLeaderElectionEventHandler.java
Comment (reswqa) [reply flink-runtime/src/test/java/org/apache/flink/runtime/leaderelection/TestingLeaderElectionEventHandler.java:79]: After looking at the relevant code again, It make sense to leave them as is.

File: flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/LeaderElectionEventHandler.java
Comment (reswqa) [inline flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/LeaderElectionEventHandler.java:62]:
```suggestion
     * <p>This method will trigger the revoke event processing in a separate thread.
```

File: flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/LeaderElectionEventHandler.java
Comment (reswqa) [inline flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/LeaderElectionEventHandler.java:85]:
```suggestion
     * <p>This method will trigger the leader information change event processing in a separate thread.
```

**Review by reswqa — APPROVED** [review]
Looks good from my side % java doc.

File: flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/LeaderElectionEventHandler.java
Comment (XComp) [reply flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/LeaderElectionEventHandler.java]: Oh man :facepalm: Thanks for being precise in your review :+1:

File: flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/LeaderElectionEventHandler.java
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/LeaderElectionEventHandler.java:50]: What's the long-term plan here; will we keep both variants and everything has to support both? Or will we drop the sync variants once we remove legacy stuff?

Comment (XComp) [reply flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/LeaderElectionEventHandler.java:50]: In the long run, we could remove the async calls again from the interface. The driver implementations only need the synchronous methods only because their event handling lives in a separate thread already. The asynchronous execution would become an implementation detail of `DefaultLeaderElectionService`

Comment (XComp) [reply flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/LeaderElectionEventHandler.java:50]: Considering that, adding deprecation annotations here would make sense.

Comment (zentol) [reply flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/LeaderElectionEventHandler.java:50]: Why can't it be an implementation detail of the `DefaultLeaderElectionService` _now_?

File: flink-runtime/src/test/java/org/apache/flink/runtime/leaderelection/DefaultLeaderElectionServiceTest.java
Comment (XComp) [reply flink-runtime/src/test/java/org/apache/flink/runtime/leaderelection/DefaultLeaderElectionServiceTest.java:191]: I was puzzled about that one as well at the start. But it makes sense: The HA backend (i.e. the driver) holds the ground truth of leadership. If the HA backend indicates the leadership loss, no operation should be executed that relies on `hasLeadership(UUID)` because another leader could have picked up in the meantime already. The `onRevokeLeadership` call that follows is used for cleaning up the component's leadership state only. Does that make sense?

I could add a description to that assert to avoid other's wonder the same when reading the test code.

File: flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/LeaderElectionEventHandler.java
Comment (XComp) [reply flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/LeaderElectionEventHandler.java:50]: How would we determine that we're running in the main thread? We would need to get a reference to the owner of `DefaultLeaderElectionService`, wouldn't we? That way we would be able to check whether the call is run in the main thread of the owner. But this would closer coupling of the owner and the `DefaultLeaderElectionService`. Or do you have something else in mind which I'm missing? :thinking: 

My understanding is that we have the following requirements:
* `onGrantLeadership`, `onRevokeLeadership` and `onLeaderInformationChanged` should be called in a single thread to ensure sequential execution.
* the methods shouldn't be called in the main thread of the `LeaderElectionService`'s owner. The k8s and ZK implementation do this implicitly through their event thread handling. The `DefaultMultipleComponentLeaderElectionService` doesn't do that (anymore). But that's an implementation detail of the service. That's why I thought that it's also the `DefaultMultipleComponentLeaderElectionService`'s responsibility to call the event processing functions asynchronously.

WDYT?

Comment (XComp) [reply flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/LeaderElectionEventHandler.java:50]: > We would need to get a reference to the owner of `DefaultLeaderElectionService`, wouldn't we?

Ok, I was not thinking straight when writing the previous comment. We have the `LeaderContender`'s reference. 🤦 I'm gonna give it some thought.

Comment (zentol) [reply flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/LeaderElectionEventHandler.java:50]: > How would we determine that we're running in the main thread?

Do we need to know that?

File: flink-runtime/src/test/java/org/apache/flink/runtime/leaderelection/DefaultLeaderElectionServiceTest.java
Comment (zentol) [reply flink-runtime/src/test/java/org/apache/flink/runtime/leaderelection/DefaultLeaderElectionServiceTest.java:191]: That makes sense; a comment would be useful.

**Comment by XComp** [comment]
I reverted the changes that were done in `DefaultMultipleComponentLeaderElectionService` because they are not necessary to make the leader event handling run in a separate thread in `DefaultLeaderElectionService` (as figured out in [the comment above](https://github.com/apache/flink/pull/22422#discussion_r1175063876). `DefaultMultipleComponentLeaderElectionService` will be removed at the end of the FLINK-26522 refactoring anyway.

I added a `@NotThreadSafe` annotation to the `LeaderElectionEventHandler` interface to acknowledge that the methods of that interface have to be called from the same thread.

**Review by zentol — APPROVED** [review]

**Comment by XComp** [comment]
Thanks for the reviews. I squashed the commits in https://github.com/apache/flink/commit/e1997df38d2bd380804a1a1814d14b75db2493ed and rebased the branch in https://github.com/apache/flink/commit/97082098f891fb0b169538ca8445cf7810e660e2

**Comment by XComp** [comment]
fixed test failure and rebased branch.

**Comment by XComp** [comment]
There's a test failure that popped up in [CI](https://dev.azure.com/apache-flink/apache-flink/_build/results?buildId=48461&view=logs&j=af184cdd-c6d8-5084-0b69-7e9c67b35f7a&t=0f3adb59-eefa-51c6-2858-3654d9e0749d&l=15266): Due to the fact that the drivers could trigger the event handling methods (i.e. `onRevokeLeadership` and `onGrantLeadership`) in a separate thread, it could happen that we stop the `DefaultLeaderElectionService` but there's still a event handling call in some thread queue that's triggered after the service is stopped (and consequently, the service's `leadershipOperationExecutor`). That will lead to a `RejectedExecutionException` because we're not checking whether the service is still running before putting the task in the executor's pool.

We actually have unit tests for covering this issue. That just didn't fire because the `DirectExecutorService` implementation doesn't follow the `ExecutorService`'s interface contract. I created FLINK-31995 to cover this issue.

**Comment by XComp** [comment]
I had to rebase because of FLINK-31962.

**Comment by XComp** [comment]
Squashed the commits together again to prepare the PR for the merge.

**Comment by XComp** [comment]
Did a final rebase after PR #22517 was merged to `master` to run the test suite with the proper `DirectExecutorService` implementation.

**Pull request merged**
